rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/nodesource/NodeSourceConfigurationParser.java

```diff
@@ -0,0 +1,140 @@
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive_grid_cloud_portal.rm.client.nodesource;
+
+import java.util.HashMap;
+
+import org.ow2.proactive_grid_cloud_portal.rm.client.NodeSourceConfiguration;
+import org.ow2.proactive_grid_cloud_portal.rm.client.PluginDescriptor;
+import org.ow2.proactive_grid_cloud_portal.rm.client.RMController;
+
+import com.google.gwt.json.client.JSONArray;
+import com.google.gwt.json.client.JSONException;
+import com.google.gwt.json.client.JSONObject;
+
+
+public class NodeSourceConfigurationParser {
+
+    private RMController controller;
+
+    public NodeSourceConfigurationParser(RMController controller) {
+        this.controller = controller;
+    }
+
+    public NodeSourceConfiguration parseNodeSourceConfiguration(String json) throws ImportException {
+
+        JSONObject jsonObject;
+
+        try {
+            jsonObject = this.controller.parseJSON(json).isObject();
+        } catch (JSONException e) {
+            throw new ImportException("The imported node source is not a valid JSON file", e);
+        }
+
+        try {
+            String nodeSourceName = jsonObject.get("nodeSourceName").isString().stringValue();
+            boolean nodesRecoverable = jsonObject.get("nodesRecoverable").isBoolean().booleanValue();
+
+            JSONObject infrastructurePluginDescriptorJson = jsonObject.get("infrastructurePluginDescriptor").isObject();
+            String infrastructurePluginName = infrastructurePluginDescriptorJson.get("pluginName")
+                                                                                .isString()
+                                                                                .stringValue();
+            PluginDescriptor infrastructurePluginDescriptor = getPluginDescriptor(infrastructurePluginDescriptorJson,
+                                                                                  infrastructurePluginName);
+
+            JSONObject policyPluginDescriptorJson = jsonObject.get("policyPluginDescriptor").isObject();
+            String policyPluginName = policyPluginDescriptorJson.get("pluginName").isString().stringValue();
+            PluginDescriptor policyPluginDescriptor = getPluginDescriptor(policyPluginDescriptorJson, policyPluginName);
+
+            return new NodeSourceConfiguration(nodeSourceName,
+                                               nodesRecoverable,
+                                               infrastructurePluginDescriptor,
+                                               policyPluginDescriptor);
+        } catch (RuntimeException e) {
+            throw new ImportException("The imported node source has incorrect parameters.", e);
+        }
+    }
+
+    public HashMap<String, PluginDescriptor> parsePluginDescriptors(String json) {
+
+        JSONArray arr = this.controller.parseJSON(json).isArray();
+        HashMap<String, PluginDescriptor> plugins = new HashMap<>();
+
+        for (int i = 0; i < arr.size(); i++) {
```

Comment:
return Stream.of(arr).map(object -> arr.get(i).isObject()).map(jsonObject -> jsonObject.get("pluginName").isString().stringValue()).map(pluginName -> getPluginDescriptor(p, pluginName)).collect(Collectors.toList());

Reply:
I have troubles applying this one.

Reply:
Ok, it is fine. I applied it, much cleaner. Thank you

Reply:
Somehow the JSONArray of the GWT library is not suitable for stream as it does not extend any collection class. It is also not iterable. I guess the for loop is the only choice left. Same remark for the next comment...